### DISCUSSION TITLE
How to a load file in a new tab from the API

### DISCUSSION THREAD

**Discussion by matteyeux** [Vector35/binaryninja-api#2066] [API] [answered]
Hi, 
is there a way to load a file in a new tab using the API ?

I'm doing `bv = BinaryViewType.get_view_of_file("my_macho")` it seems to load and analyse it but it doesn't open a new tab.

For the context, I'm writing a loader for Apple's sep-firmware which is a file that contains multiple Mach-O files. I was able to find and extract them. But I have no idea how to load a single file in a new tab or window once it's extracted.

In the end I would like the user to choose which part of the firmware to extract and load in Binja (see picture below).

![image](https://user-images.githubusercontent.com/8758978/97336553-70f5c400-187f-11eb-9745-89cf10e9451b.png)

Thanks.

**Comment by psifertex** [comment, marked as answer]
So first an answer to your specific question in the title, and then I'll follow up with a better solution for your actual problem (thanks for explaining both, gets you a two-for-one!).

To open a new file (if you're already running code from a UI Plugin you can short-circuit much of this if you have a UIContext already):
```python
execute_on_main_thread_and_wait(lambda: UIContext.activeContext().openFilename('/tmp/test.bndb')
```

You could also do something terrible like: 
```python
from PySide2.QtGui import QDesktopServices
QDesktopServices.openUrl("binaryninja:///bin/ls")
```

However, for an even cleaner workflow, you can instead create your own view settings that let the user select possible views from within the settings UI and then just only create the BinaryView they select/have your code just return the bytes from the one view selected (that's how we handle FAT Mach-O files currently).

However, I'll need to tag-in one of the other developers to include that answer, so I'll link them to this thread and let them explain the right way to do that!

**Comment by matteyeux** [comment]
Jordan, 

Opening a file with `execute_on_main_thread_and_wait` seems to be the exact thing I wanted to do thanks ! 

Is there any doc for creating my own view settings ? Is this related to the [settings module](https://api.binary.ninja/binaryninja.settings-module.html) ?

Looking forward the best way to do this then :)

**Reply by psifertex** [reply]
So it turns out I was incorrect. The other workflow I thought was possible isn't, so the approaches above are the right way to go for now. If you keep an eye on #133  whenever that is resolved you'll be able to support those more natively.
